[SE-0283] Implement Equatable, Comparable, and Hashable conformance for Tuples

Azoy opened the pull request:
From: https://forums.swift.org/t/hacking-equatable-conformance-on-void/25975

The tests right now are kind of lackluster, what else can I be testing with this? What else can be improved upon in this implementation? Is there an edge case that I hadn't thought about?

cc: @jckarter @rjmccall

Azoy commented on stdlib/public/runtime/KnownProtocolWitnessTables.cpp:
```diff
@@ -0,0 +1,68 @@
+}
+
+const _ProtocolConformanceDescriptor swift::
+BUILTIN_PROTOCOL_CONFORMANCE_DESCRIPTOR_SYM(EMPTY_TUPLE_MANGLING,
```
Currently this triggers a global constructor warning. I imagine it's because I reference the variable within the constructor? Is there any way to fix this?

rjmccall replied:
This kind of pointer subtraction is not officially a constant expression.  Compilers can support it, but most won't emit it without a global constructor.

I think it would be much more reasonable to find a way to trick the compiler into emitting this.

Azoy commented:
@rjmccall I've updated the implementation to have IRGen emit conformance descriptors for the stdlib. Let me know if this is what you had in mind.

rjmccall commented on include/swift/AST/ProtocolConformance.h:
```diff
@@ -1015,6 +1021,89 @@ class InheritedProtocolConformance : public ProtocolConformance,
 };
 
+/// A builtin conformance appears when a special non-nominal type has a runtime
+/// declared conformance. E.g. the runtime implements equatable for all Void.
```
Capitalize `Equatable`, please.  And just "Void", not "all Void".

rjmccall commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -2734,9 +2734,18 @@ IRGenModule::getAddrOfLLVMVariable(LinkEntity entity,
+    // address point, but ()'s metadata is the full metadata type, so the
+    // element bit cast is wrong for both address types.
+    if (entity.isTypeMetadata(Context.TheEmptyTupleType))
+      return existing;
```
I feel like the special cases here are definitely being added at the wrong level.

Azoy replied:
How would you suggest go about this? If I ask for LinkEntity's FullMetadata address point, I get the correct global variable here, but it mangles to Mf vs. N. I could alter LinkEntity's mangling to check for these types if that makes more sense.

rjmccall replied:
Well, let's see if this is actually necessary.  If this becomes a general conformance on tuple types, we'll need to be able to record that kind of specially, and the usefulness of directly linking to metadata will go away.

rjmccall commented on lib/SIL/TypeLowering.cpp:
```diff
@@ -1424,7 +1425,24 @@ static CanTupleType computeLoweredTupleType(TypeConverter &tc,
+          isOpaqueEquatableVoid = true;
+      }
+    }
+  }
```
This should just be added to `matchesTuple`; the `::Type` cases should allow `OpaqueTypeArchetypeType`s.  This hasn't been an issue before because tuples haven't had any protocol conformances.

rjmccall reviewed (commented):
Yes, this is basically what I was asking for, thank you.

As a general matter, I wonder if it wouldn't be better to just go a step further and make this a conditional conformance for all tuple types.  That would be a big request except that you've obviously already figured out how to do some quite sophisticated stuff.

Azoy commented:
I'll look into doing this conditionally for all tuples to see what work is needed.

dan-zheng commented on include/swift/Runtime/KnownProtocolWitnessTables.h:
```diff
@@ -0,0 +1,78 @@
+///
+/// Note: Empty tuples are not lowered as parameters, so it's fine to exclude.
+SWIFT_RUNTIME_EXPORT SWIFT_CC(swift)
+bool BUILTIN_PROTOCOL_WITNESS_SYM(EMPTY_TUPLE_MANGLING,
```
I'm curious if this approach is a step towards enabling tuple conformances to arbitrary protocols? Or is it strictly limited to tuple conformances to specific known protocols?

Azoy replied:
This approach essentially allows us to create conformances for any structural type to any protocol. We're only limiting this to certain protocols.

If you're referring to something like extending tuples and allowing them to declare conformances there, then there's more work needed there. I don't think tuple extensions are possible without parameterized extensions, but you'd also need variadic generics for those extensions to be really useful anyway. You know, the *usual*. x)

dan-zheng replied:
Thanks for the response!

> This approach essentially allows us to create conformances for any structural type to any protocol. We're only limiting this to certain protocols.

I see. It seems to me that "*known* protocol witness tables" is inherently an approach for hardcoding conformances, so it's not user-extensible.

> If you're referring to something like extending tuples and allowing them to declare conformances there, then there's more work needed there. I don't think tuple extensions are possible without parameterized extensions, but you'd also need variadic generics for those extensions to be really useful anyway.

Yes, declaring tuple conformances in source code was my original question, sorry for not being more clear. Parameterized extensions make sense as an enabler:
```swift
extension<T: Equatable, U: Equatable> (T, U): Equatable { ... }
```

The ability to declare tuple conformances seems useful even without variadic generics: at least it becomes possible to express tuple conformances!

Azoy replied:
I'm unsure whether this patch is a step towards that. I imagine once we have a real implementation of tuple extensions that we'd want to treat them as normal protocol conformances and emit actual witness table implementations. I did some reading on previous discussions of this and some were under the impression that these builtin conformances would be considered technical debt once we get a proper implementation, which I somewhat agree with. However, I'm also under the impression that doing this now is probably better than waiting for parameterized extensions (I'm ashamed I haven't worked on it in a while) and variadic generics. I can imagine that the work to get actual tuple conformances working after those land *shouldn't* be totally difficult.

jckarter commented on include/swift/ABI/MetadataValues.h:
```diff
@@ -361,10 +361,14 @@ enum class TypeReferenceKind : unsigned {
 
   First_Kind = DirectTypeDescriptor,
-  Last_Kind = IndirectObjCClass,
+  Last_Kind = DirectTypeMetadata,
```
If you were going to generalize this to all tuples, you might have this new type reference kind instead be a reference to a mangled name in the generic context of the conformance. That should allow any kind of type to be instantiated without hardcoding type-specific metadata access patterns.

rjmccall replied:
Yeah.  If we're preparing for arbitrary structural conformances in this PR, there's no sense in adding the ability to directly link to metadata when that's only likely to ever be useful for the  empty tuple.  This could link to a mangled name, although that might unnecessarily duplicate the information in the conformance; alternatively, it could "link" to a `MetadataKind`.

Azoy commented:
@rjmccall @jckarter I've gone ahead and implemented Equatable conformance for all tuples. Let me know what other tests you would like and any improvements I can make to the implementation. This is pretty exciting!

Azoy commented:
Currently this initializes a new witness table everytime we need one, but it should be possible to have a single witness table in the runtime. Hmm...

jckarter commented:
The conformance descriptor format isn't absolutely important. Fundamentally, we just need something the witness table can point to that uniquely identifies the conformance. It seems to me like we should be able to have one descriptor and one witness table that works for all tuples.

Azoy commented:
@jckarter Sorry I was being dumb, does this look ok? Single descriptor and witness table now!

jckarter commented:
I haven't had time yet to look at this PR in depth, but I hope to soon. Sorry!

Azoy commented:
@jckarter I've *hopefully* added backward compatibility for these conformances for both 5.0 and 5.1 runtimes. I'm not too versed with how the compatibility library works, but let me know if I made a mistake or something of the like. The test for compatibility should be exercised through the interpreter test I added in earlier commits, correct? I remember reading something from here: https://github.com/apple/swift/pull/25030#discussion_r287166816

Azoy commented on stdlib/toolchain/Compatibility50/BuiltinProtocolConformances.cpp:
```diff
@@ -0,0 +1,92 @@
+    // If we don't have a conformance then something somewhere messed up in
+    // deciding that this tuple type was Equatable...??
+    if (!conformance)
+      abort();
```
`fatalError` seems to be missing from these compatibility libraries?

jckarter replied:
The compatibility libraries might in some cases be linked into a binary that doesn't itself use `libswiftCore`, such as a main app binary that is all C/ObjC/C++ but which transitively uses some library containing Swift that needs the compatibility binary, so it can't directly use Swift runtime facilities. For something like `fatalError`, adding a local version that logs a message and calls `abort` should be fine. (More generally, you might need to "soft link" Swift runtime functionality by using dlopen/dlsym.)

jckarter commented on lib/IRGen/GenProto.cpp:
```diff
@@ -3396,3 +3402,54 @@ llvm::Constant *IRGenModule::getAddrOfGenericEnvironment(
+
+    addProtocolConformance(std::move(description));
+  }
+}
```
Could these data structures be implemented in the runtime C++? It might be easier for people to find and update their implementation that way. (You might have to resort to inline assembly to deal with the relative references though…)

Azoy replied:
They were originally in the runtime until John recommended I move the conformance descriptor into the compiler to remove relative reference issues. Inline assembly is a clever way around that however. I'll look into it because this would clean up a bit of this implementation (this would let me directly reference the witness table now.)

Azoy replied:
Would this require defining the whole structure in inline assembly? Because this might make it a lot less portable and confusing rather than redirecting people to this irgen file to update or add new conformance descriptors.

jckarter replied:
Since we require clang to build the runtime already, the inline assembly format for data should be portable. Another possibility would be to write it as LLVM IR.

Azoy replied:
I believe that I've completed this, but there may be some issues with how I've set it up. I know for a fact that what I have now doesn't work on Windows, but I don't know how to correctly test it. Would love if you took a look and let me know what I need to change about this.

jckarter commented on lib/IRGen/GenProto.cpp:
```diff
@@ -1792,8 +1792,15 @@ namespace {
+      // Builtin conformance type references are just MetadataKind, which
+      // require no relative reference.
+      if (isa<BuiltinProtocolConformance>(Conformance)) {
+        B.add(ref.getValue());
```
It isn't clear what the comment means here. Is `ref.getValue()` the constant `MetadataKind`, or is it still referring to the metadata record?

jckarter replied:
(This could just be residual confusion from me looking at the old `()` specific version of the patch though. In context, it's clear that it's the `MetadataKind`.)

Azoy replied:
My apologizes if the comment wasn't so clear, hopefully if we move the descriptors to the runtime this just falls out.

jckarter commented on stdlib/toolchain/Compatibility50/BuiltinProtocolConformances.cpp:
```diff
@@ -0,0 +1,98 @@
+//===--- BuiltinProtocolWitnessTables.cpp ---------------------------------===//
```
Any time we link against `Compatibility50`, we should also link against `Compatibility51`, since we'd be backward deploying to support both 5.0 and 5.1 runtimes at that point. Instead of duplicating this code in both compatibility libraries, I think you should be able to put it once in the `Compatibility51` library, and refer to the `51` symbols from within `50`'s hooks table.

Azoy replied:
Oh that's interesting. I'll try and see if I can get this working.

Azoy replied:
> refer to the `51` symbols from within `50`'s hooks table.

It seems to me that this is referring to replacing the `swift50override_conformsToProtocol` in `50`'s hook table with `51`'s, but that doesn't seem right because it doesn't use any of the backported bug fixes for `50`. I assume I'm interpreting this incorrectly, but does it make sense to call `51`'s implementation within `50`'s implementation?

jckarter replied:
Yeah, in the case of conformsToProtocol where there are 5.0-specific fixes, we should have the swift50override perform those fixes then chain into the 5.1 version.

jckarter commented on include/swift/Runtime/BuiltinProtocolConformances.h:
```diff
@@ -0,0 +1,82 @@
+#define SWIFT_EQUAL_OPERATOR_MANGLING 2eeoiySbx_xtFZ
+
+// (A...)
+#define VARIADIC_TUPLE_MANGLING xd_t
```
For now, would this mangling only be used for the witness symbols?

Azoy replied:
Yes, this mangling would only be used for the witness symbols. I imagine variadic generics would get their own mangling separate from a variadic generic type, but if they aren't then this might pose an issue.

jckarter replied:
As long as it's only for the symbols, I think we're OK, as long as we note that this mangling sequence is used for that purpose. It's not hard to hard code existing symbol names of things for ABI stability purposes if we do establish a more generic variadic mangling in the future.

Another possibility might be to just give these symbols fixed unmangled names like `_swift_tupleEquatableWitnesses`, `_swift_tupleEquatable_equals`, etc. that the compiler knows.

Azoy replied:
I've gone ahead and just used the unmangled names to play it safe just in case that mangling is ever needed in the future.

futurejones commented:
Can one of the admins verify this patch?

Azoy commented:
Does it make sense to also implement comparable and hashable in this pr, or should those be in separate prs?

jckarter commented:
@Azoy It might be easier to review to break them out into their own PRs. Thanks!

jckarter commented on include/swift/Runtime/BuiltinProtocolConformances.h:
```diff
@@ -0,0 +1,56 @@
+// MachO requires a leading underscore for symbols.
+#if defined(__MACH__)
+#define TUPLE_EQUATABLE_CONF "__swift_tupleEquatable_conf"
+#define TUPLE_EQUATABLE_WT "__swift_tupleEquatable_wt"
```
You can replace the conditional here with the predefined `__USER_LABEL_PREFIX__` macro:

```
#define TUPLE_EQUATABLE_CONF __USER_LABEL_PREFIX__ "_swift_tupleEquatable_conf"

jckarter commented on lib/AST/Module.cpp:
```diff
@@ -864,6 +864,23 @@ ProtocolConformanceRef ModuleDecl::lookupConformance(Type type,
+    if (protocol == ctx.getProtocol(KnownProtocolKind::Equatable)) {
+      // Ensure that every element in this tuple conforms to Equatable.
+      for (auto eltTy : tuple->getElementTypes()) {
+        auto conformance = lookupConformance(eltTy, protocol);
```
Sorry I missed this earlier, but we should really save the element conformances and use them as arguments to the builtin conformance. At runtime, we should similarly instantiate the witness table pattern with the collected witness tables for all of the element conformances. This will be more efficient, because we save having to dynamically look up those conformances, and will be robust in the case when there are multiple retroactive `Equatable`/`Comparable`/`Hashable` conformances for the same type in a process.

Azoy replied:
I think I'm confused because you said earlier in this thread that it seems reasonable to have a single descriptor and witness table for all tuples. Is the instantiated witness table pattern separate from the witness table?

Also, when you mean arguments to the builtin conformance, are you suggesting arguments when we're instantiating the builtin conformance? (In hindsight this does make sense because they'll be readily accessible in SIL when we can devirtualize some witness calls.)

jckarter replied:
Sorry for the confusion. We can still have a single descriptor, and the implementations of the protocol witness methods can remain constant, but the witness table itself would need to be instantiated with the conformances for all of the tuple elements, so that the implementation can call into them. For an analogy to what needs to happen, you could look at what happens when you declare a conditional conformance on a nominal type:

```
struct Pair<T, U> {}

extension Pair: Equatable where T: Equatable, U: Equatable { }

func foo<T: Equatable>(_: T.Type) {}

foo(Pair<Int, String>.self)
```

Because `Pair` itself does not require T or U to be Equatable, the type metadata does not have any record of the conformance. The `T: Equatable` and `U: Equatable` conformances will be collected and passed to instantiate the `Pair<T, U>` conformance when we call `foo`.

Azoy commented:
@jckarter Okay, I believe that should do it. Would love another round of review!

Azoy commented:
I just noticed that while the equals witness doesn't call `conformsToProtocol` anymore, when we call `conformsToProtocol` with a tuple type I still check that each element conforms to said protocol. Maybe there's a more principled way to indicate that a tuple conforms to `Equatable`?

jckarter commented:
I think that's the right approach. You'd have to call `conformsToProtocol` to gather the witness tables for each of the elements in order to instantiate the witness table for the tuple's conformance.

jckarter commented:
Re: the static witness table version in the compatibility libraries, it looks like I'm going to have to prise out more of the conformance cache logic into the Compatibility51 library in order to backport the fix from https://github.com/apple/swift/pull/29899 to older runtimes. Maybe that'll make it easier to have uniform behavior in the back-deployed version.

Azoy committed:
[AST] Introduce BuiltinProtocolConformance

Azoy committed:
[ABI] Introduce MetadataKind TypeRef

Azoy commented:
Hey @jckarter , in regards to the new compatibility library for 5.1, how should this handle adding fixes to 5.2? I assume another compat library is needed for 5.2, but does the protocol conformance cache move into 5.2 aswell? Or do I implement the fixes in the 5.1 library and simply call that implementation from the 5.2 compat library? Is this going into 5.3? Do I need to make 2 more compat libraries? Sorry for the bombardment of questions!

jckarter commented:
Scheduling-wise, I would assume this isn't going to make it into 5.3. We can generally assume that, if a `CompatibilityXX` library is linked, all of the later compatibility libraries are also linked (`Compatibility50` implies `51`, `52`, etc.), so we can move logic into the latest compatibility library where it's needed. If you were able to implement the backward-deployment handling on top, without modifying the conformance cache, it might still be better to stick with that implementation approach, and have your tuple handling chain back into `swift_conformsToProtocol`, to reduce the chance of disruption. I only tested the conformance cache when running against 5.0 and 5.1 runtimes, and there's a chance that it might've drifted from the 5.2 or 5.3 logic in the weeks since. We may only need to add a `Compatibility53` library if there's no 5.2-specific backward deployment code.

Azoy committed:
Implement Tuple Equatable Conformance

Azoy commented:
@jckarter Hi Joe, I’m running into an issue where some of the assembly structures in the compatibility libraries are causing an issue with linking in `Compatibility50Linking.c` test. These structures reference some of the swift symbols that that test doesn’t link against causing the issue. I’ve seen the 51 library does some dlsym with swift symbols to retrieve functions or structures, which I’ve also resorted to, but can’t do for the assembly structures. Sorry there hasn’t been much movement on this in a while (new job!), but I do have all of the conflicts resolved and have reorganized the commit history, just have to push. Thanks so much!

Azoy commented:
Ok, so I think I know of a solution to remove the hard dependency on libswiftCore, but it'll require that the conformance descriptor in the compatibility library not reference a protocol at all. This'll either require one of two solutions: 1. that `swift_getWitnessTable` be a compatibility function in 53 to directly hand the pre made witness table for the compatibility libraries or 2. have a different lowering strategy in IRGen depending on what runtime version of swift we're deploying. If we're doing the second option, the current lowering is having the conditional requirements and witness table generation go through `swift_getWitnessTable`, but for the compatibility libraries we'll instead just reference the pre made witness table directly. I'm unsure which solution is best, or if what I'm saying doesn't make any sense whatsoever (most likely case), but I would love some input so we can move forward with this. 

I messed around with weak linking the symbol in the conformance descriptor, but it still required that the symbol be resolved by another linked dependency vs. being undefined like elf's. (I marked it as both `.weak_reference` and `.weak_definition`, but no luck.)

Azoy committed:
[Compatibility53] Add compatibility library for 5.3 and backport tuple Equatable conformance

Fix some comments

Unnecessary cast

Azoy commented:
I believe I've correctly implemented the compatibility hacks for this (with the help of @gwynne !) Please let me know what I can change/make better or if there are any other tests I should make.

gwynne commented on stdlib/toolchain/Compatibility53/BuiltinProtocolConformances.cpp:
```diff
@@ -0,0 +1,136 @@
+  TUPLE_EQUATABLE_CONF ":\n"
+  // This is an indirectable relative reference to the Equatable protocol
+  // descriptor. However, this is 0 here because the compatibility libraries
+  // can't have a dependency on libswiftCore (which is where Equatable live).
```
```suggestion
  // can't have a dependency on libswiftCore (which is where Equatable lives).
```

gwynne commented on stdlib/toolchain/Compatibility53/BuiltinProtocolConformances.cpp:
```diff
@@ -0,0 +1,136 @@
+
+// Due to the fact that the compatibility libraries can't have a hard
+// dependency to libswiftCore (which is where the Equatable protocol desciptor
+// live), we have to manually implant this before calling any user code.
```
```suggestion
// lives), we have to manually implant this before calling any user code.
```

gwynne commented on stdlib/toolchain/Compatibility53/BuiltinProtocolConformances.cpp:
```diff
@@ -0,0 +1,136 @@
+// dependency to libswiftCore (which is where the Equatable protocol desciptor
+// live), we have to manually implant this before calling any user code.
+__attribute__((constructor))
+void _implaceEquatableDescriptor() {
```
`implace` is not a word. Use `implant` or `emplace`.
```suggestion
void _emplaceEquatableDescriptor() {
```

gwynne commented on stdlib/toolchain/Compatibility53/BuiltinProtocolConformances.cpp:
```diff
@@ -0,0 +1,136 @@
+// live), we have to manually implant this before calling any user code.
+__attribute__((constructor))
+void _implaceEquatableDescriptor() {
+  auto tupleEquatableConf = const_cast<int32_t *>(
+    reinterpret_cast<const int32_t *>(&_swift_tupleEquatable_conf));
```
For extremely pedantic correctness, instead of `int32_t *`, `decltype(reinterpret_cast<ProtocolConformanceDescriptor *>(nullptr_t).Protocol)::Offset`. Which, I mean, won't work, even if my syntax was correct (which it isn't, my C++ is way out of practice), but it's gotta be possible to at least figure out the appropriate `Offset` type for `InProcess::RelativeIndirectablePointer`, I'd think...

gwynne replied:
Or maybe better yet (emphasis on maybe), `const_cast<std::remove_const_t<reinterpret_cast<ProtocolConformanceDescriptor const *> > >(&_swift_tupleEquatable_conf)`, then `tupleEquatableConf->Protocol = ...`, which also won't work because the ivar is private, among a half dozen other reasons... What I'm getting at regardless is see if you can make the types at least a little more generic 🤷‍♀️. If it even matters.

gwynne commented on stdlib/toolchain/Compatibility53/BuiltinProtocolConformances.cpp:
```diff
@@ -0,0 +1,136 @@
+    // If we don't have a conformance then something somewhere messed up in
+    // deciding that this tuple type was Equatable...??
+    if (!conformance)
+      abort();
```
Is there a version of something like `_internalInvariant()` for compatibility libs where you can at least include a message? Like `llvm_unreachable()` (ref: https://llvm.org/docs/CodingStandards.html#assert-liberally) if the Swift compiler headers don't have something more Swift-runtime-specific?

theblixguy replied:
I think you can call `crash` or `fatalError`, which is defined [here](https://github.com/apple/swift/blob/ad0a8d31217455bb8878708897a44b1496290cdb/include/swift/Runtime/Debug.h).

gwynne commented:
@swift-ci please smoke test

Azoy commented:
@xedin @hborla (sorry for the mention!) I found an interesting bug while implementing Comparable conformance for tuples.
```swift
func compare<T: Comparable>(_ lhs: T, _ rhs: T) -> Bool {
  lhs < rhs
}

// tuples conform if their elements conform
// bool does not conform, but this gets past the type checker
// and results in a compiler crash later.
_ = compare((false, false), (false, false))

let x = false
_ = compare((x, x), (x, x)) // error (Bool, Bool) does not conform
```
I ran this with `-debug-constraints` and it seems the CS system is giving me a bunch of type variables for conformance checking (type variables do conform to protocols). I assume there's a lot of checking for `ExpressibleByBooleanLiteral`, but the CS system doesn't seem to check the concrete types until after it's found the solution, in which case it doesn't care that it doesn't conform. It's been a while since I've messed with the CS system, any advice would be grateful!

Azoy committed:
Implement Tuple Comparable Conformance

Add protocol witnesses for all Comparable requirements

Azoy committed:
[Compatibility53] Backport Tuple Comparable Conformance

Add protocol witnesses for Comparable requirements

remove unused table

Azoy committed:
[Runtime] Tuple Comparable associated conformance is not evenly aligned

Azoy committed:
Implement Tuple Hashable Conformance

Azoy committed:
[Compatibility53] Backport Tuple Hashable Conformance

Azoy commented:
cc: @jckarter @milseman I've updated this PR to include all 3 protocol conformances per Michael's request. Hopefully I've cleaned up the commit history to make it easier to digest in review. Please let me know what else I can do. This should be ready for initial testing. Thank you so much @jckarter for helping me along the way!

jckarter commented on stdlib/public/runtime/BuiltinProtocolConformances.cpp:
```diff
@@ -0,0 +1,810 @@
+  "  .long (got." EQUATABLE_DESCRIPTOR_SYMBOL " - \
+              (" TUPLE_EQUATABLE_CONF ")) + 1\n"
+  #elif defined(__MACH__)
+  "  .long " EQUATABLE_DESCRIPTOR_SYMBOL "@GOTPCREL + 5\n"
```
We should be able to use a direct reference to EQUATABLE_DESCRIPTOR_SYMBOL without going through the GOT, since it is in the standard library and the runtime and standard library are always linked together. Same with the other symbols you generate GOT equivalents for here. You can also use `.` to refer to the current address in assembly language instead of doing the offset math yourself. So this could be:

```
TUPLE_EQUATABLE_CONF ":\n"
"  .long " EQUATABLE_DESCRIPTOR_SYMBOL " - .\n"
```

And that should also be portable between ELF and Mach-O targets as well.

If you do end up needing GOT entries, though, IIRC the syntax in ELF for a GOTPCREL reference is `symbolname@GOT`; you shouldn't need to make your own GOT equivalent variable.

Azoy replied:
That's what I thought too when I initially wrote this, but stumbled upon `'symbol' can not be undefined in a subtraction expression` for the stdlib symbols.

Regarding ELF `@GOT`: Interesting! Any reason LLVM doesn't emit this for linux binaries? I know LLVM is smart enough to use the Darwin syntax when it encounters those GOT entries.

As I'm compiling this after rebasing from upstream, it seems the build-script also wants to build the arm64 stdlib on my x86_64 mac, which leads to a linker error (of course!) `illegal text-relocation to 'symbol'` for the `==` method descriptor reference (I'm curious why it complains for that symbol and not the protocol descriptor?) Any idea how I can fix this, or debug on my machine? After some very basic research, maybe it's referring to absolute addressing on arm64 not being allowed? Although if that were the case then the protocol descriptor would error out too... I can only assume it's also going to error out with the other method descriptors as well.

Azoy replied:
Looking at llvm source, it looks like aarch64 (aka arm64, right?) doesn’t support GOT pc rel annotation, but it does support `foo@GOT - .`... I’ll do some more investigating once I’m off work.

Azoy replied:
I did some reading, and it turns out ELF has `@GOTPCREL` as well! It differs from macho in that the required `+ 4` is not required, so I can do `symbol@GOTPCREL + 1` for an indirect indirectable pointer. On ARM, it looks like `@GOT - .` is the equivalent from what I've read (which also compiles), but I haven't been able to test it yet. Even though these symbols require going through GOT, I can still take advantage of cleaning up the math by using `- .` in places where GOT is not needed. Please take a look at the latest commit and let me know what you think!

jckarter replied:
Ah, right, LLVM needs an upstream fix for x86-64 to allow for PC-relative references to undefined symbols. This should be OK then.

jckarter commented on stdlib/public/runtime/BuiltinProtocolConformances.cpp:
```diff
@@ -0,0 +1,810 @@
+                                             Metadata *conformingType,
+                                             void **witnessTable) {
+  auto tuple = cast<TupleTypeMetadata>(assocType);
+  std::vector<void *> instantiationArgs;
```
We could `reserve` capacity for the instantiationArgs here, to reduce vector growth overhead.

jckarter reviewed (approved):
Overall, this is great! Just a few small things I noticed:

gwynne commented on include/swift/AST/ProtocolConformance.h:
```diff
@@ -1013,6 +1019,106 @@ class InheritedProtocolConformance : public ProtocolConformance,
+  /// Determine whether the witness for the given requirement
+  /// is either the default definition or was otherwise deduced.
+  bool usesDefaultDefinition(AssociatedTypeDecl *requirement) const {
+    llvm_unreachable("builtin-conformances never have associated types");
```
Certainly wouldn't mind seeing, say, `ExpressibleBy<Scalar>Literal` on single-element or zero-or-one-nonoptional-element tuples... or even better, `ExpressibleByArrayLiteral` for homogenous tuples... like, for example, the ones `ClangImporter` tends to offer up? (Make them `Collection`s too, while I'm at it!) Synthesized `CaseIterable` in a future where cases of enums with associated values can be referred to as if lacking said values... (appropriate respects to @tdotclare for the idea of working with enums in such a fashion, even if I have my doubts!)

Only other builtin protocol I can think of offhand with an associated type would be `Identifiable`... I suppose I could see that being, ahem, "interesting" if synthesized on closure types...

Azoy committed:
[ABI] Split TypeReferenceKind from TypeMetadataRecord

Azoy committed:
[Runtime] Clean up the assembly around Tuple builtin conformances

forgot + 1

jckarter commented:
@swift-ci Please test

Azoy commented:
The linux failure is due to some missing `\n` on the associated conformance, but I don't really understand the iOS failure here. It looks like llvm crashed here compiling `_swift_tupleHashable_hashValue` for arm64e-apple-ios7.0.0. I recall hearing something about having to sign function pointers if ptrauth is enabled, perhaps that's the issue here? I'll try building this target and playing around to see if I can fix this.

Azoy committed:
[Runtime] Add stubs for indirect symbols on armv7/s, add missing newlines for ELF directives, fix compile crash for arm64e

Azoy commented:
@jckarter I've added the missing newlines for ELF, so linux tests should pass now. I've built the iphoneos targets and fixed the compile crash for arm64e, but ran into an issue with armv7/s with their indirect symbols, so I had to add stubs for non_lazy_symbol_ptr's for them. I can't test that these arch changes work (because x86_64 can only run executable tests for x86_64 and I think arm64 for iphoneos simulator). I had a fun time learning about how llvm handles these archs for things like `@GOT` syntax, which Darwin doesn't support for armv7/s!.

jckarter commented:
Yeah, armv7 is restricted in how it can form relative references because the Mach-O format doesn't have all the necessary relocations to do it in the general case, so if you have to do something hacky to get it working, that's fine. Thanks for working through it!

jckarter commented:
@swift-ci Please test

Azoy commented:
Let me just build all the Apple targets... and run the validation suite as well...

Azoy committed:
[Runtime] Emit non lazy stub pointers for 32 bit x86 for tuple conformances

jckarter commented:
Looks like there are a couple conflicts that crept in:

```
Conflicting files
lib/Serialization/ModuleFormat.h
test/stdlib/Compatibility50Linking.c
```

@Azoy do you mind clearing those up, and I'll kick off PR testing again?

Azoy commented:
@jckarter done!

benrimmington commented:
@swift-ci Please test

benrimmington commented:
```
--- check-swift-all-linux-x86_64 ---
PASS: Swift(linux-x86_64) :: stdlib/Tuple.swift.gyb (7553 of 13585)
PASS: Swift(linux-x86_64) :: Interpreter/builtin_conformances.swift (11317 of 13585)

--- check-swift-all-optimize-linux-x86_64 ---
FAIL: Swift(linux-x86_64) :: stdlib/Tuple.swift.gyb (7553 of 13585)
FAIL: Swift(linux-x86_64) :: Interpreter/builtin_conformances.swift (11314 of 13585)

--- check-swift-validation-macosx-x86_64 ---
PASS: Swift(macosx-x86_64) :: Interpreter/builtin_conformances.swift (11257 of 13512)
PASS: Swift(macosx-x86_64) :: IRGen/builtin_conformances.swift (12101 of 13512)

--- check-swift-validation-iphonesimulator-i386 ---
FAIL: Swift(iphonesimulator-i386) :: Interpreter/builtin_conformances.swift (11258 of 13512)
FAIL: Swift(iphonesimulator-i386) :: IRGen/builtin_conformances.swift (12109 of 13512)
```

Azoy committed:
Manually define _rawHashValue in compatibility libraries, AST fixes for builtin protocol conformance

Remember to use substituted type in builtin conformance substitution

Azoy commented:
@jckarter I have fixed the issues that arose when building those tests with optimizations and have ensured that they pass locally. So, Linux should be good to go. As for the i386 iphone simulator, I was unable to run tests because we currently spawn an `iPhone 5` instance and none of the newer runtimes support that device, so I couldn't run that specific device. However, I did build and was able to test on watchos simulator i386 which had the same issue as the iphone simulator. I've tested locally that these changes fix the listed failures that @benrimmington posted, so hopefully we should be go to go this time around!

xwu commented:
@swift-ci Please test

benrimmington commented:
@Azoy The instructions for creating an "iPhone 5" simulator configuration (on Intel-based Macs) are in [docs/Testing.md][].

[docs/Testing.md]: <https://github.com/apple/swift/blob/master/docs/Testing.md#the-llvm-lit-based-testsuite>

benrimmington commented:
* The test failure in stdlib/Tuple.swift.gyb (check-swift-all-optimize-linux-x86_64) looks similar to last time.

  ```
  SIL verification failed: concrete type lookup requires conformance that matches type
  ```

  Can you use the same options as `--preset=buildbot_linux_1604` to test locally?

* The other failure is new, but I don't think the Windows platform was tested last time.

  ```
  S:\jenkins\workspace\swift-PR-windows\swift\stdlib\public\runtime\BuiltinProtocolConformances.cpp(106,14): error: expected ')'
    "  .long " INDIRECT_RELREF_GOTPCREL(EQUATABLE_DESCRIPTOR_SYMBOL) "\n"
               ^
  S:\jenkins\workspace\swift-PR-windows\swift\stdlib\public\runtime\BuiltinProtocolConformances.cpp(89,6): note: to match this '('
  __asm(
       ^
  S:\jenkins\workspace\swift-PR-windows\swift\stdlib\public\runtime\BuiltinProtocolConformances.cpp(262,14): error: expected ')'
    "  .long " INDIRECT_RELREF_GOTPCREL(COMPARABLE_DESCRIPTOR_SYMBOL) "\n"
               ^
  S:\jenkins\workspace\swift-PR-windows\swift\stdlib\public\runtime\BuiltinProtocolConformances.cpp(245,6): note: to match this '('
  __asm(
       ^
  S:\jenkins\workspace\swift-PR-windows\swift\stdlib\public\runtime\BuiltinProtocolConformances.cpp(624,14): error: expected ')'
    "  .long " INDIRECT_RELREF_GOTPCREL(HASHABLE_DESCRIPTOR_SYMBOL) "\n"
               ^
  S:\jenkins\workspace\swift-PR-windows\swift\stdlib\public\runtime\BuiltinProtocolConformances.cpp(607,6): note: to match this '('
  __asm(
       ^
  3 errors generated.
  ```

* Should we also "test source compatibility" before merging?

Azoy commented:
@benrimmington yes, sorry this Linux failure got through. I’m able to reproduce and am working on a solution. Yes, the windows failure is expected because I don’t currently have support for that. I remember @milseman wanting to run source compatibility before merging, so I believe we should test that too.

Azoy committed:
[Runtime] Add Native Windows support for builtin conformances

Force C name mangling for Windows

Azoy commented:
@benrimmington I've tested using the linux preset that both of those failing tests now pass. I've also added Windows support for these conformances, so we can start testing Windows as well (I haven't tested the Windows support yet, but I am comparing what I wrote with some Swift assembly generated on Windows). It might be a good idea to go ahead and start source compatibility tests (assuming all tests pass). Sorry all for taking so long to try and fix these bugs, but I appreciate everyone sticking with me!

theblixguy commented:
@swift-ci please test

theblixguy commented:
@swift-ci please test source compatibility

theblixguy commented:
macOS failure:

> 22:01:07 dyld: Symbol not found: __swift_tupleEquatable_conf
22:01:07   Referenced from: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/lib/swift/macosx/libswiftSwiftPrivate.dylib
22:01:07   Expected in: /usr/lib/swift/libswiftCore.dylib

theblixguy commented:
@swift-ci please test Windows

Azoy commented:
The Windows failure looks like an easy fix, but I'm really confused how macOS failed this time around (even though I didn't change anything with its symbol??). Even more confusing that macOS failed, but both source compatibility tests runs succeeded.

Azoy commented:
It looks like some of these other built libraries (`libswiftSwiftPrivate` in the test, but `libswiftDarwin` for me) are looking for `__swift_tupleEquatable_conf` and friends in the system runtime vs. looking for it in the built runtime (at least `objdump` is telling me they are linking against `@rpath/libswiftCore.dylib`). I assume this is intentional, but I don't know the options how to fix this. I'm curious how the previous macOS test passed whereas this one failed now.

jckarter commented:
That doesn't seem right to me. I would expect the locally-built library dylibs to link against the locally-built runtime.

Azoy commented:
Yeah, it doesn't seem right to me either, but the macOS failure that @theblixguy posted earlier and the failures I'm seeing locally state that the symbol was
> Expected in: /usr/lib/swift/libswiftCore.dylib

jckarter commented:
Yeah, that seems like a build system bug to me. The other dylibs ought to be looking at the just-built runtime.

Azoy commented:
Here I've just fixed the Windows issue along with resolving some of the merge conflicts. I'm still seeing the macOS failure happening, but I think it's because the `unittests/runtime` test is using the system `libswiftCore.dylib`... I'm still looking into it.

kastiglione commented:
@Azoy have you reproduced the error locally?

Azoy commented:
@kastiglione Yes, I have. I'm getting some help from others to try and figure out this issue!

Azoy commented:
@compnerd helped me uncover the issue, and it seems to be a deep rooted build system issue. I'm going to try and patch a fix so that this can be unblocked.

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

Azoy commented:
Now that https://github.com/apple/swift/pull/34302 has been pushed, this should be good to go!

benrimmington commented:
@swift-ci Please test

benrimmington commented:
@swift-ci Please test Windows platform

benrimmington commented:
@swift-ci Please test source compatibility

benrimmington commented:
Errors due to `swift_runtime_unreachable` — renamed to [`swift_unreachable`](https://github.com/apple/swift/blob/main/include/swift/Basic/Unreachable.h).

Azoy commented:
Yea I've already changed it locally. Going to wait for the other tests to run to see if there are any other issues I should fix before pushing again.

benrimmington commented:
(Succeeded on Linux. Failed on Windows with link error.)

> StdlibUnittest.obj : error LNK2019: unresolved external symbol __imp__swift_tupleEquatable_conf referenced in function $s14StdlibUnittest11expectEqual___10stackTrace9showFrame4file4lineySS_SsSSyXKAA14SourceLocStackVSbSSSutF

jckarter commented:
Do we mark the symbol as `dllimport` when targeting PE platforms?

Azoy commented:
I don’t mark these symbols as anything on PE platforms right now. I thought marking the assembly as `globl` would suffice in exporting this symbol. I did read somewhere on some msvc docs that symbols that start with `__imp_` are JMP stubs whereas `__imp__` prefixed symbols are IAT entries, so I’m thinking that it’s just unfortunate that these symbols have leading underscores causing the IAT entry instead of a JMP stub (I could most likely be wrong).

jckarter commented:
@Azoy Yeah, I think @compnerd set it up to add `dllimport/dllexport` automatically, but you might need to use the right abstraction to do so. Maybe he can help if you can't find it yourself.

Azoy commented:
@benrimmington I would love another round of testing! I've built Swift on Windows locally and appear to have fixed the issue, I've also gone ahead and renamed my uses of `swift_runtime_unreachable` to `swift_unreachable`.

benrimmington commented:
@swift-ci Please test

benrimmington commented:
@swift-ci Please test Windows platform

Azoy commented:
I've just realized that the IRGen test is going to fail on Windows because I forgot to conditionalize `dllimport` on the conformance symbol, I apologize!

benrimmington commented:
```
S:\jenkins\workspace\swift-PR-windows\swift\test\IRGen\builtin_conformances.swift:3:17: 
error: CHECK-LABEL: expected string not found in input

// CHECK-LABEL: @_swift_tupleEquatable_conf = external global %swift.protocol_conformance_descriptor

                ^
```

```
<stdin>:55:1: note: possible intended match here

@_swift_tupleEquatable_conf = external dllimport global %swift.protocol_conformance_descriptor, align 4

^
```

Azoy commented:
@benrimmington alright, this should be the one! I've pushed the fix for the test to conditionalize the dllimport.

benrimmington commented:
@swift-ci Please test

benrimmington commented:
@swift-ci Please test Windows platform

jckarter commented:
@Azoy Have you asked for commit access yet? If you have commit access then you ought to be able to trigger CI by yourself.

benrimmington commented:
All checks have passed. Should we test source compatibility again?

Azoy commented:
I would say yes because they failed last time due to the macOS failure.

benrimmington commented:
@swift-ci Please test source compatibility

xwu commented:
@Azoy, this is fantastic. I guess it's ready to go?

Azoy commented:
@xwu I believe so! @jckarter are we clear for takeoff?

jckarter commented:
Should be good to go. Thanks @Azoy !

drodriguez commented:
Has an internal crash in clang appeared during development?

It seems to have started around the introduction of these changes. The full logs can be seen in https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android-arm64/6614/console.

The failed assertion seems to be

```
clang++: /home/ubuntu/jenkins/workspace/oss-swift-RA-linux-ubuntu-16.04-android-arm64/llvm-project/llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFObjectWriter.cpp:120: virtual unsigned int (anonymous namespace)::AArch64ELFObjectWriter::getRelocType(llvm::MCContext &, const llvm::MCValue &, const llvm::MCFixup &, bool) const: Assertion `(!Target.getSymA() || Target.getSymA()->getKind() == MCSymbolRefExpr::VK_None || Target.getSymA()->getKind() == MCSymbolRefExpr::VK_PLT) && "Should only be expression-level modifiers here"' failed.
```

In the case of the Android ARM32, the error is slightly different and might give someone a clue:

```
<inline asm>:1:41: error: expected STT_<TYPE_IN_UPPER_CASE>, '#<type>', '%<type>' or "<type>"
  .type __swift_tupleEquatable_private, @object
                                        ^
<inline asm>:5:37: error: expected STT_<TYPE_IN_UPPER_CASE>, '#<type>', '%<type>' or "<type>"
  .type _swift_tupleEquatable_conf, @object
                                    ^
<inline asm>:23:58: error: expected STT_<TYPE_IN_UPPER_CASE>, '#<type>', '%<type>' or "<type>"
  .type "associated conformance _swift_tupleComparable", @object
                                                         ^
<inline asm>:33:42: error: expected STT_<TYPE_IN_UPPER_CASE>, '#<type>', '%<type>' or "<type>"
  .type __swift_tupleComparable_private, @object
                                         ^
<inline asm>:37:38: error: expected STT_<TYPE_IN_UPPER_CASE>, '#<type>', '%<type>' or "<type>"
  .type _swift_tupleComparable_conf, @object
                                     ^
<inline asm>:62:40: error: expected STT_<TYPE_IN_UPPER_CASE>, '#<type>', '%<type>' or "<type>"
  .type __swift_tupleHashable_private, @object
                                       ^
<inline asm>:66:36: error: expected STT_<TYPE_IN_UPPER_CASE>, '#<type>', '%<type>' or "<type>"
  .type _swift_tupleHashable_conf, @object
```

There's some `__asm` pieces in the code, but I cannot match the reported errors to any of them.

Does someone has any idea what might be going on?

drodriguez commented:
I found the problematic assembly. It was collapsed by GitHub.

Seems that `@` is a comment symbol in AArch32 according to https://reviews.llvm.org/D10651. `__ELF__` is probably exercised only in Linux, and not in ARM. ARM has probably checked in `__MACH__`, but not others.

Does anybody know which symbol can be used in front of `object` so it works both in every ELF platform (or at least in x86 and ARM)?

Azoy commented:
@drodriguez for ARM32 on Android, it looks like it wants `%object` which should be a quick fix, but the ARM64 android looks like it doesn't support `@GOTPCREL` relocation from the error you gave (either nothing or `@PLT` which is different iirc). @jckarter I think a lot of these issues could be avoided if we manually created got equivalent structures for linux platforms, but that'll increase the amount of assembly for this dramatically. Maybe we can move this to an `.S`?

drodriguez commented:
Thanks for the answer. I found that the Linux AArch64 testing seems to also had failed with a similar error: https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-18.04-aarch64/2177/console. I will try to double check tonight about using the percent sign instead of the at-symbol.

jckarter commented:
@Azoy The ELF syntax is `@GOT` instead of `@GOTPCREL`, IIRC.

aschwaighofer commented:
Looks likely this broke: https://ci.swift.org/job/oss-swift_tools-RA_stdlib-DA_test-device-non_executable/250/

aschwaighofer commented:
Temporary disabled the test here: https://github.com/apple/swift/pull/34480

aschwaighofer commented:
@Azoy Can you take a look, I expect that the test just needs updating.

tbkka commented:
We're also seeing some tests fail with errors about invalid relative offsets in a conformance.  I don't know for sure that it was this PR, but the timing is suspicious.  I'm trying to get more details.

tbkka commented:
@rjmccall @jckarter -- Do you see a way that @azoy could declare these conformances without resorting to assembly?

Azoy commented:
@tbkka we're discussing writing them in LLVM IR instead over here: https://github.com/apple/swift/pull/34457

jckarter commented:
@Azoy would it be OK if you reverted this? Another issue we're seeing is that it appears that executables and dylibs that link the compatibility versions of the tuple conformance implementation end up reexporting the `__swift_tuple*` symbols. The compatibility implementation ought to be private to the binary it gets linked in, so we'll want to make sure they have `hidden` visibility in the compatibility .a files (and it'd probably be good to give them different symbol names like `__swift50_tuple*` so they don't collide with the real runtime implementations in newer OSes).

I wonder if the issues @tbkka alluded to about invalid conformance addresses might be related to the compatibility libraries as well—if we link the compatibility libraries into a binary, but run with a new runtime that contains the real implementation of tuple conformance, that real implementation needs to take precedence, and the compatibility implementation would be invalid from the host runtime's perspective. Binaries that relies on the compatibility library for backward deployment might need to use a stub function that picks either the compatibility or host runtime version of the symbols to resolve the address of these symbols, instead of referencing the symbols directly.

airspeedswift commented:
Reverted in #34492
